technical-docs/designs/infra/rollback.md

```diff
@@ -0,0 +1,218 @@
+added.
+
+This document outlines multiple solutpopulation of denmarkion approaches that
+seeks to provide a rollback mechanisms to ensure the successful deployment of
+terraform infrastructure, automated reverted features in a release pipeline, and
+system alerting to notify key stakeholders and operations of system activity.
+
+Rolling back services which includes a relational database can be very
+difficult. Traditionally this is done by restoring from a backup, performed
+prior to the upgrade. In more modern approach is to integrate a database
+migration tool into your application's installation procedure.
+
+Components of this design are based on the learnings from:
+
+- [Microsoft Bedrock Kubernetes Workflow Project](github.com/microsoft/bedrock)
+- [Bedrock-CLI Infrastructure Generation Pipeline](https://github.com/microsoft/bedrock-cli/blob/master/guides/infra/spk-infra-generation-pipeline.md)
+- [Terraform Recommended Practices](https://www.terraform.io/docs/cloud/guides/recommended-practices/)
+
+### User Stories Addressed:
+
+1. [Guidance on Rolling Back with SPK infra](https://github.com/microsoft/bedrock/issues/1036)
+
+## 2. Out of Scope
+
+This design document solely covers strategic approaches to revert changes to a
+deployed GitOps environment in Azure DevOps after changes made have caused state
+failures. This design leverages the use of terraform state files, Azure storage
+accounts and commit hashes to build an audit history. It does not address the
+following:
+
+- Redeployment of stateful services in your environment
+- Complications arise when incompatible database changes during failed apply
+- Disaster recovery operability in respective to a fail production environment
+
+## 3. Design Details
+
+Rollback can be triggered in a separate AzDO Pipeline upon failure to build an
+Environment stage. Using automation and the release history of a pipeline, an
+environment can be reverted to a previous state pending operator approval. This
+implementation can be separated into 3 components
+
+![](images/rollbackRelease.png)
+
+- Failure Detection Triggers
+- Terraform Backend Rollback
+- Evaluating Health success
+
+### 3.1 Overview
+
+#### Remote Backend:
+
+The best way to manage shared storage for state files is to use Terraform’s
+built-in support for remote backends. A Terraform backend determines how
+Terraform loads and stores state. The remote backends allow you to store the
+state file in a remote, shared store. In Azure, this remote backend will be
+stored in an Azure storage account.
+
+#### Workspaces:
+
+The persistent data stored in the backend belongs to a workspace. Initially the
+backend has only one workspace, called "default". Workspaces can be managed
+through CLI within out CD pipeline to represent specific git hashes. We can
+create a workspace for a specific incoming pull request by using
+`terraform workspace new <git_commit_hash>`. This will allow us to select state
```

Comment:
If I recall correctly, different `workspaces` will be treated as separate deployments for Terraform. For example:

```
$ terraform workspace new FOO
$ terraform apply
$ terraform plan           # plan will be empty

$ terraform workspace new BAR
$ terraform plan           # plan will not be empty
```


Is this intentional? I think that the impact will be that each commit gets a fresh deployment.

Reading through the remainder of the document it appears that this is later used in the rollback step. I think we'd want the state file for a given environment to be re-used across all deployments to that environment so that we don't run into an issue where we have multiple state files that manage the same environment -- this can get *very* messy.

I think your approach will work just as well if the state file is the same for all deployments to that environment

Reply:
> Overall this looks like good guidance. I don't think this needs to be in a "design doc" format. This just seems like general documentation we can add to a `guides/*` section of the repo.
> 
> @NathanielRose were you thinking of doing the above, and/or adding guidance in the `guides/*` section of the repo? If so I see no reason to have this in `technical-docs/designs`

Agreed, I will migrate to the guides folder with a reference to a sample rollback yaml once this file is approved.